Sources/SwiftParser/Lexer/Cursor.swift

```diff
@@ -133,11 +133,61 @@ extension Lexer.Cursor {
       case .afterStringLiteral: return false
       case .afterClosingStringQuote: return false
       case .inStringInterpolationStart: return false
-      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _, stringInterpolationStart: _): return stringLiteralKind != .multiLine
+      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _): return stringLiteralKind != .multiLine
       case .afterStringInterpolation: return false
       }
     }
   }
+
+  /// A data structure that holds the state stack entries in the lexer. It is
+  /// optimized for situations where there's at most one state on the stack
+  /// (in addition to the bottom `normal` mode) and to not create any ARC
+  /// traffic.
+  ///
+  /// It does this by representing the bottom `normal` state implicitly and
+  /// storing one additional state inline. If the stack needs to contain more
+  /// entries, the required memory is allocated in a bump allocator that is
+  /// expected to outlive the stack.
+  struct StateStack {
+    private var topState: State? = nil
+    private var stateStack: UnsafeBufferPointer<State>? = nil
+
+    var currentState: State {
+      return topState ?? .normal
+    }
+
+    mutating func perform(stateTransition: Lexer.StateTransition, stateAllocator: BumpPtrAllocator) {
+      switch stateTransition {
+      case .push(newState: let newState):
+        if let topState = topState {
+          if let stateStack = stateStack {
+            let newStateStack = stateAllocator.allocate(State.self, count: stateStack.count + 1)
+            let (_, existingStateStackEndIndex) = newStateStack.initialize(from: stateStack)
+            newStateStack[existingStateStackEndIndex] = topState
```

Comment:
I don't think `topState` should be in `newStateStack`.
for top-normal: `topState=nil`, `stateStack=nil`
for top-some: `topState=.some`, `stateStack=nil`
for non-top: `topState=.some`, `stateStack=[...]`

Reply:
That’s exactly what we do (or I’m completely misunderstanding what you’re trying to say). If you don’t count the bottom `normal` state as an entry in the stack size we have

- Stack size 0: `topState=nil`, `stateStack=nil`
- Stack size 1: `topState=.some`, `stateStack=nil`
- Stack size >1 `topState=.some`, `stateStack=[…]`

If we are in this `if`, then we have stack size >=1 (since `topState != nil`). So what we do is to transfer `topState` into `newStateStack` so that we can assign `newState` to `topState` (which now lives in `stateStack`)

Reply:
Sorry I misread your PR. Ignore this.